Add methods to get argument count of methods

AThousandShips opened the pull request:
Added to:
* `Callable`s
* `Object`
* `ClassDB`
* `Script(Instance)`s

Todo:
- [x] Document
- [x] Add tests
- [x] Implement for C# (needs confirmation)
- [x] Implement for GDExtension (confirmed)
- [x] Implement for godot-cpp (needs confirmation)

Help with how to do this in C# would be very welcome, I have some ideas but I'm not very familiar with the C# internals

To confirm the godot-cpp side, see:
* https://github.com/godotengine/godot-cpp/pull/1375

Might split off the extension part into a separate PR if desired

---

* Closes: https://github.com/godotengine/godot-proposals/issues/8596

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

AThousandShips committed:
Add methods to get argument count of methods

Added to:
* `Callable`s
* `Object`s
* `ClassDB`
* `Script(Instance)`s

AThousandShips commented on core/object/object.h:
```diff
@@ -856,6 +856,7 @@ class Object {
 	Variant property_get_revert(const StringName &p_name) const;
 
 	bool has_method(const StringName &p_method) const;
+	int get_method_argument_count(const StringName &p_method) const;
```
Unsure if I should add an equivalent method to `ClassDB`

AThousandShips commented:
Changing so that when there's no defined number, i.e. when the Callable is invalid, or when there's no such method, it returns -1 to distinguish, will push soon

AThousandShips commented:
Unsure if the various methods should error instead of silently returning `0`, but leaving as is right now

AThousandShips commented on modules/mono/managed_callable.cpp:
Unfamiliar with C# and if this needs any specific treatment here or if I should just leave it alone

AThousandShips commented on modules/mono/signal_awaiter_utils.cpp:
Same here

AThousandShips commented on core/variant/callable.h:
```diff
@@ -153,6 +154,7 @@ class CallableCustom {
 	virtual void call(const Variant **p_arguments, int p_argcount, Variant &r_return_value, Callable::CallError &r_call_error) const = 0;
 	virtual Error rpc(int p_peer_id, const Variant **p_arguments, int p_argcount, Callable::CallError &r_call_error) const;
 	virtual const Callable *get_base_comparator() const;
+	virtual int get_argument_count() const = 0;
```
Keeping this virtual for now, but once I've identified the use cases I will probably make it non-virtual, just wanting to make sure I don't miss anything anywhere

AThousandShips commented:
Tested the Extension side in godot-cpp, will open a PR for that too but can confirm it works with the relevant changes

AThousandShips commented:
I believe the GDExtension part is ready, but would appreciate some feedback on that

CC @godotengine/gdextension 

Ditto for C# side

CC @godotengine/dotnet

AThousandShips commented:
Lot of details to work out but ready for review by the relevant teams, will add tests maybe tomorrow, at least in GDScript

Mickeon commented:
This may be the first time function arguments are acknowledged with method names. It has me pondering if this should be the standard across the whole class reference ("parameter" vs "argument")

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -682,6 +682,14 @@
+			<return type="int" />
+			<param index="0" name="method" type="StringName" />
+			<description>
+				Returns the number of arguments of the given [param method].
```
Consistency with `has_method`.
```suggestion
				Returns the number of arguments of the given [param method] by name.
```

AThousandShips commented:
Not the first time, see `get_bound_arguments_count` etc.

Mickeon commented on doc/classes/ClassDB.xml:
```diff
@@ -65,6 +65,15 @@
+			<param index="1" name="method" type="StringName" />
+			<param index="2" name="no_inheritance" type="bool" default="false" />
+			<description>
+				Returns the number of arguments of the method [param method] of [param class] or its ancestry if [param no_inheritance] is [code]false[/code].
```
So many "of"s here. Perhaps replacing one would still make sense?
```suggestion
				Returns the number of arguments for the method [param method] of [param class] or its ancestry if [param no_inheritance] is [code]false[/code].
```

AThousandShips replied:
I'm not so sure `for` is grammatical here, it is the arguments *of* the method, but will see what I can do with the wording

Mickeon commented on doc/classes/Callable.xml:
```diff
@@ -112,6 +112,12 @@
+		<method name="get_argument_count" qualifiers="const">
+			<return type="int" />
+			<description>
+				Returns the total number of arguments of this [Callable], accounting for any bound (or unbound) arguments. The number of bound arguments is subtracted from the total, and the number of unbound arguments are added.
```
I understand what this is trying to say but it is a bit confusing when worded like this. I have no suggestions at the moment.

AThousandShips replied:
Me too, was trying to figure out how to say it

Mickeon replied:
Okay, I'll help you think. From the outsider developer's perspective, what changes? What needs to be known? What will this method be used for?

Judging by the code, I would say the important part to get here is that what we mean by "arguments" is "bunch of values that _can_ (oftentimes _should_) be passed into the Callable"
```suggestion
				Returns the total number of arguments this [Callable] should take, including optional arguments.
```
The part about `bind` and `unbind` is a consequence of this definition.

AThousandShips replied:
I like that, I'd keep the bind/unbind explicit as it can be very, *very* confusing (I got it backwards the first time I wrote the code and added instead of subtracted), but I'll take that suggestion as a base 🙂 thank you!

Mickeon replied:
Oh by all means, it needs to be specified to remove any doubts. It's just that -- and I repeat myself -- by rewording the definition this way it makes the behavior of `bind` and `unbind` look like a natural consequence of it, instead of a potentially confusing exception to the norm.

Bromeon commented on core/extension/gdextension_interface.h:
```diff
@@ -420,6 +422,8 @@ typedef struct {
 	GDExtensionCallableCustomLessThan less_than_func;
 
 	GDExtensionCallableCustomToString to_string_func;
+
+	GDExtensionCallableCustomGetArgumentCount get_argument_count_func;
 } GDExtensionCallableCustomInfo;
```
I think this is a breaking change:

If someone running a 4.2 extension declares the struct on the binding side, it won't have that last field. If this extension is then run in Godot 4.3, Godot will try to access `get_argument_count_func`, causing UB.

Unfortunately it might need a `GDExtensionCallableCustomInfo2` iteration, but this also impacts other functions 😕 ideas?

AThousandShips replied:
Then I'll leave the the extension side to future implementation perhaps, wasn't sure hence asking, will look at those details but will split it off to keep it simple on this side 🙂 thank you

Bromeon replied:
If there are other similar additions in the future (e.g. advanced argument queries for type/default-ness), it could make sense to batch some changes in a minor release, to avoid proliferation of struct iterations.

AThousandShips replied:
Indeed, will open a new PR to discuss this more coherently and make a new structure with the related changes there and update the godot-cpp PR, will link here when done

Bromeon replied:
I see you added a `GDExtensionScriptInstanceInfo3`, but is this single niche feature really worth another iteration of the struct -- which means extra compat code on both Godot + binding side?

I would really consider holding out until we can pack more into this; otherwise we'll end up with `GDExtensionScriptInstanceInfo7` soon 😉

AThousandShips replied:
Thinking a bit further I might restrict it to the script side and not on the instance side, we'll see!

dsnopek replied:
I have an unrelated change that I've been wanting to make to `GDExtensionScriptInstanceInfo2` - I can try to make a PR for it for 4.3 so this isn't the only change to that struct :-)

AThousandShips replied:
Then I'll keep this in 🙂

akien-mga replied:
Let's try to get @dsnopek's change in with this one for 4.3 then, or keep both (and potential future additions) for 4.4 maybe?

AThousandShips replied:
Agreed, both at the same time, keeping the extension code in its separate commit until we decide and alternatively splitting it off

dsnopek replied:
@akien-mga @AThousandShips Here's a PR with the change I had in mind: https://github.com/godotengine/godot/pull/89055

AThousandShips commented:
Re-added the GDExtension code, with new compatibility details, works well in my testing (but haven't tested the ScriptInstance extension part as godot-cpp doesn't utilize it)

AThousandShips commented on core/extension/gdextension_interface.h:
```diff
@@ -2312,6 +2392,7 @@ typedef GDExtensionScriptInstancePtr (*GDExtensionInterfaceScriptInstanceCreate)
 /**
  * @name script_instance_create2
  * @since 4.2
+ * @deprecated in Godot 4.3. Use `script_instance_create3` instead.
```
Forgot to update the line in the `script_instance_create` part, will do later

AThousandShips commented on core/object/script_language_extension.h:
```diff
@@ -101,6 +101,8 @@ class ScriptExtension : public Script {
 	EXBIND1RC(bool, has_method, const StringName &)
 	EXBIND1RC(bool, has_static_method, const StringName &)
 
+	EXBIND1RC(int, get_method_argument_count, const StringName &)
```
Could remove this part from the extension and rely on `get_method_info` but it might not provide all the necessary details, but it would make extending easier and only affect the `Callable` part

Just as an option to reduce the extent of extension changes

raulsntos commented on modules/mono/managed_callable.cpp:
```diff
@@ -85,6 +85,11 @@ ObjectID ManagedCallable::get_object() const {
 }
 
+int ManagedCallable::get_argument_count() const {
+	// TODO: I'm not sure if this is needed as I don't know where this is used
+	return 0;
```
This would be called when you call `get_argument_count` on a `Callable` created from a C# delegate/lambda. You'd need to create a method in `DelegateUtils.cs` like the `DelegateEquals` and `DelegateHash` methods because we can only retrieve the delegate information from C#.

AThousandShips replied:
Good, then I will see how that can work out, should be workable

AThousandShips replied:
I've solved this one, and discovered that the signal awaiters are accessible from the GDScript side with `get_connections` on a `Signal` so trying out some code, will push soon

AThousandShips commented:
Added basic tests both for GDScript and in tests

Can look at adding more general callable tests in the future but out of scope for this

AThousandShips commented on modules/mono/signal_awaiter_utils.cpp:
```diff
@@ -90,6 +90,11 @@ ObjectID SignalAwaiterCallable::get_object() const {
 }
 
+int SignalAwaiterCallable::get_argument_count() const {
+	// TODO: I'm not sure if this is needed as I don't know where this is used
+	return 0;
```
Then I'll just have them defaulted 🙂 Thank you!

AThousandShips commented:
The C# side should be complete now, I'll leave the event signal handlers alone as internal details

Will work some more on the documentation tomorrow or the day after

Waiting for some more feedback from C# and GDExtension on the specifics here

AThousandShips commented on modules/mono/signal_awaiter_utils.h:
Might make these warn when calling them and indicate they're undefined, alternatively add a warning to `CallableCustom` noting it's not implemented for this case

AThousandShips commented:
Reorganized the code, added validity return data for all the checks in `Object`, `ClassDB`, `Script`, and `ScriptInstance`, added a fallback method to the instances and separated the instance extension into its own commit for ease of integration

AThousandShips commented on doc/classes/ScriptExtension.xml:
```diff
@@ -80,6 +80,12 @@
 			<description>
 			</description>
 		</method>
+		<method name="_get_script_method_argument_count" qualifiers="virtual const">
```
This now expects either an `int` if valid, or `null` if not, will see about documenting but this file is sparse already so will see

KoBeWi replied:
`Return the expected argument count for the given [param method], or [code]null[/code] if it can't be determined.`

I've seen new methods added here many times, but they never get a description because "the class is not documented anyway".

AThousandShips replied:
I'll document it in the extension case, it doesn't have it doesn't actually handle `null` special rn, will look though

AThousandShips replied:
I'll add it tomorrow, or in the follow up PR, someone (not me) should document all of these though

AThousandShips commented on modules/mono/csharp_script.cpp:
```diff
@@ -1792,6 +1792,34 @@ bool CSharpInstance::has_method(const StringName &p_method) const {
 			gchandle.get_intptr(), &p_method);
 }
 
+int CSharpInstance::get_method_argument_count(const StringName &p_method, bool *r_is_valid) const {
```
Will look at writing an instance bridge method for this like the `has_method` part, will see if needed

AThousandShips commented on core/object/object.cpp:
```diff
@@ -679,6 +679,58 @@ bool Object::has_method(const StringName &p_method) const {
+
+	{
+		bool valid = false;
+		int ret = ClassDB::get_method_argument_count(get_class_name(), p_method, &valid);
```
Now delegates to the `ClassDB` case instead, keeping the code in one place

AThousandShips commented on core/object/script_instance.cpp:
```diff
@@ -32,6 +32,28 @@
 
 #include "core/object/script_language.h"
 
+int ScriptInstance::get_method_argument_count(const StringName &p_method, bool *r_is_valid) const {
```
Simple delegation here, very much like how it's done in GDScript but without the dedicated features used there, allows for using this feature without extending `ScriptInstance`, though it might not catch all cases, and might not be as optimal

AThousandShips commented on core/object/script_language.cpp:
```diff
@@ -102,6 +102,22 @@ Dictionary Script::_get_script_constant_map() {
 	return ret;
 }
 
+int Script::get_script_method_argument_count(const StringName &p_method, bool *r_is_valid) const {
```
Fallback in the same way, this means that as long as an extended script has `get_method_info` with appropriate return information you can utilize the whole chain of functionality, so this should work out of the box for existing extensions that provide method information

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/DelegateUtils.cs:
```diff
@@ -44,6 +44,34 @@ internal static int DelegateHash(IntPtr delegateGCHandle)
+        {
+            try
+            {
+                var @delegate = (Delegate?)GCHandle.FromIntPtr(delegateGCHandle).Target;
+                if (@delegate is null)
+                {
+                    *outIsValid = godot_bool.False;
+                    return 0;
+                }
+                var @method = @delegate.Method;
+                if (@method is null)
+                {
+                    *outIsValid = godot_bool.False;
+                    return 0;
+                }
+                *outIsValid = godot_bool.True;
+                return @method.GetParameters().Length;
```
Just use the null conditional operator and we don't have to repeat the null check return.

```suggestion
                var @delegate = (Delegate?)GCHandle.FromIntPtr(delegateGCHandle).Target;
                int? argCount = @delegate?.Method?.GetParameters().Length;
                if (argCount is null)
                {
                    *outIsValid = godot_bool.False;
                    return 0;
                }
                *outIsValid = godot_bool.True;
                return argCount.Value;
```

AThousandShips replied:
Thank you! Never used these features (and haven't used C# to any real degree in over 10 years)

Will do

AThousandShips commented:
Will look at more tests later, but there's no real tests for a lot of the methods of `ClassDB` and `Object` so unsure how to go about that, so keeping it to tests for the `Callable`s for now

AThousandShips commented on core/object/script_language_extension.h:
```diff
@@ -101,6 +101,19 @@ class ScriptExtension : public Script {
+			return (ret.get_type() == Variant::INT) ? ret.operator int() : 0;
+		}
+		// Fallback to default.
+		return Script::get_script_method_argument_count(p_method, r_is_valid);
```
Unsure if it should fall back when the method doesn't return a valid value here instead, like:
```cpp
		Variant ret;
		if (GDVIRTUAL_CALL(_get_script_method_argument_count, p_method, ret) && ret.get_type() == Variant::INT) {
			if (r_is_valid) {
				*r_is_valid = true;
			}
			return ret.operator int();
		}
		// Fallback to default.
		return Script::get_script_method_argument_count(p_method, r_is_valid);
```
It'd allow falling back, but also custom behaviour, but unsure

dsnopek replied:
I'm not sure I understand. The alternative in your comment is to allow returning values that aren't integers but that could be converted to integers?

AThousandShips replied:
No it's to fall back to the default if it doesn't return an `INT`

AThousandShips replied:
Then I'll add it to my next push

AThousandShips commented on core/variant/callable.h:
```diff
@@ -153,6 +154,7 @@ class CallableCustom {
 	virtual void call(const Variant **p_arguments, int p_argcount, Variant &r_return_value, Callable::CallError &r_call_error) const = 0;
 	virtual Error rpc(int p_peer_id, const Variant **p_arguments, int p_argcount, Callable::CallError &r_call_error) const;
 	virtual const Callable *get_base_comparator() const;
+	virtual int get_argument_count(bool &r_is_valid) const;
```
Using reference here to avoid unnecessary null checks

AThousandShips replied:
Might go back to having this be virtual, but define a default behaviour to refer back to, that way new code doesn't forget to add it

AThousandShips commented on core/object/script_language.h:
```diff
@@ -150,6 +150,8 @@ class Script : public Resource {
 	virtual bool has_method(const StringName &p_method) const = 0;
 	virtual bool has_static_method(const StringName &p_method) const { return false; }
 
+	virtual int get_script_method_argument_count(const StringName &p_method, bool *r_is_valid = nullptr) const;
```
Named it this way as per the note for `has_method` above

dalexeev commented on modules/gdscript/gdscript_utility_callable.cpp:
```diff
@@ -80,6 +80,20 @@ ObjectID GDScriptUtilityCallable::get_object() const {
+		case TYPE_GDSCRIPT:
+			r_is_valid = true;
+			return GDScriptUtilityFunctions::get_function_argument_count(function_name);
+		default:
+			r_is_valid = false;
+			return 0;
```
I would prefer an explicit `case TYPE_INVALID` instead of `default` for consistency. If we potentially add a new type and forget to add the new case here, the compiler will produce a warning that is treated as an error.

AThousandShips replied:
Will need a default path with a warning to accomplish that unfortunately, but will add that, it complains of non-returning path with just the three cases

dsnopek reviewed (approved):
The GDExtension changes look good to me! I tested with godot-cpp, and everything seems to be working fine.

(I haven't reviewed the core, GDScript or C# changes - only GDExtension)

dalexeev commented on core/object/object.cpp:
```diff
@@ -679,6 +679,58 @@ bool Object::has_method(const StringName &p_method) const {
+		}
+	}
+
+	const Script *scr = Object::cast_to<Script>(this);
+	if (scr != nullptr) {
+		bool valid = false;
+		int ret = scr->get_script_method_argument_count(p_method, &valid);
+		if (valid) {
+			if (r_is_valid) {
+				*r_is_valid = true;
+			}
+			return ret;
+		}
+	}
```
Shouldn't inheritance be checked here since `Script::get_script_method_argument_count()` doesn't take into account base classes, unlike `ScriptInstance::get_method_argument_count()`? (This is quite confusing. See also #80418.)

AThousandShips replied:
Good point, I based this on how methods like `get_method_info` and `has_method` work in GDScript and Mono for the Script and ScriptInstance cases respectively, but open to other suggestions, can make this one iterate without trouble 🙂

AThousandShips replied:
Also realized this is called on the class itself, got it confused with the static part in the case above, so should probably use `script` instead of casting itself, unsure though

Will add both for comparison

AThousandShips commented:
Fixed some of the above pointers, squashed the basic extension things into the first commit (keeping the instance side separate for easier reworking) and did some other minor rework

dalexeev commented on modules/gdscript/gdscript_utility_callable.cpp:
```diff
@@ -80,6 +80,22 @@ ObjectID GDScriptUtilityCallable::get_object() const {
+		case TYPE_GDSCRIPT:
+			r_is_valid = true;
+			return GDScriptUtilityFunctions::get_function_argument_count(function_name);
+		default:
+			ERR_FAIL_V_MSG(0, "Invalid type.");
+	}
```
```suggestion
	}
	ERR_FAIL_V_MSG(0, "Invalid type.");
```

This will compile ans not suppress the warning if a case is accidentally missed.

AThousandShips commented on core/object/object.cpp:
```diff
@@ -679,6 +679,59 @@ bool Object::has_method(const StringName &p_method) const {
+		}
+	}
+
+	const Script *scr = Object::cast_to<Script>(this);
+	while (scr != nullptr) {
+		bool valid = false;
+		int ret = scr->get_script_method_argument_count(p_method, &valid);
+		if (valid) {
+			if (r_is_valid) {
+				*r_is_valid = true;
+			}
+			return ret;
+		}
+		scr = scr->get_base_script().ptr();
+	}
```
Just calling it on the script itself and letting `ScriptInstance` handle the attached script as is done elsewhere

akien-mga commented:
So this adds `GDExtensionScriptInstanceInfo3` for GDExtension. I'm not opposed to it, but it's worth considering whether we have other changes in the pipeline (implemented or planned) that would also need to change that struct, so we can bundle them all in the same release and avoid adding new structs all the time, making GDExtension full of backward compatibility structs. CC @godotengine/gdextension

AThousandShips commented:
Went ahead and split off the script instance part and will open a dedicated PR for it to make the review easier so we can merge this safely without timing it

akien-mga commented:
I think this could now be squashed and merged?

AThousandShips commented:
As long as the C# side looks good I think we're good, I'll squash tomorrow if there's no questions

AThousandShips commented:
I'll squash it up soon and we can give it a spin, in my testing the c# code worked and was based on suggestions by raulsntos, so I feel confident in going forward, if there's any questions I can always split the C# off and leave the placeholders

raulsntos reviewed (approved):
The C# changes look good to me, thanks!

KoBeWi reviewed (approved):
The implementation looks fine overall.

akien-mga commented:
Thanks!

AThousandShips commented:
Thank you! Will improve the docs and fixup the followup PR tomorrow
